---
output:
  html_document:
    toc: true                               # table of contents
    toc_float: true                         # float the table of contents to the left of the main document content
    toc_depth: 3                            # header levels 1,2,3
    theme: default
    number_sections: true                   # add section numbering to headers
    df_print: paged                         # tables are printed as an html table with support for pagination over rows and columns
    css: ./corp-styles.css
    highlight: pygments
#   code_folding: none
#   theme: spacelab
    pdf_document: true
#   self_contained: no
bibliography: ./references.bibtex
params:
  path_proj_summary: ""
  path_versions: ""
  path_design: ""
  path_config: ""
  path_contrasts: ""
  path_wd: ""
  path_genelist: ""
# Author: Simon Heumos
# Contributor: Silvia Morini, Gisela Gabernet
---

```{r, bootstrap.show.message=FALSE, echo=FALSE, message=FALSE}
df <- read.csv(file = params$path_proj_summary, sep="\t", header = TRUE)
# Import the .tsv file downloadable from https://portal.qbic.uni-tuebingen.de/portal/web/qbic/user-database -> Projects
projectcode = as.character(df$Sub.Project)
title = as.character(df$Short.Title)
PI = as.character(df$Principal.Investigator)
PI_address = as.character(df$PI.Address)
PI_email = as.character(df$PI.E.Mail)
PI_institute = as.character(df$PI.Institute)
PI_organization = as.character(df$PI.Organization)
PM = as.character(df$Project.Manager)
PM_email = as.character(df$Manager.E.Mail)
description = as.character(df$Description)
```

<!-- load config.yaml + design.txt -->
```{r load_config, bootstrap.show.message=FALSE, echo=FALSE, message=FALSE}
library(yaml)
config <- read_yaml(params$path_config)
design_deseq2 <- paste(readLines(params$path_design), collapse=" ")
```

---
title: "`r projectcode` - Project Report"
subtitle: "`r title`"
author: "`r PM`, Bioinformatics Research Scientist"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
---

<img src="./logo.png" style="position:absolute;top:0px;right:0px;" height="200" width="200" />
<div class="watermark">QBiC</div>

```{r docSetup, bootstrap.show.code = FALSE, dev = 'png', bootstrap.show.message=FALSE, echo=FALSE, message=FALSE}
## knitrBoostrap and device chunk options
options(warn = -1)                          # suppress warnings and messages
library(knitr)
opts_chunk$set(bootstrap.show.code = FALSE, dev = 'png', echo=FALSE)
```

\
\
\
\

**Project Members:** 

***`r PI`***

`r PI_email`

`r PI_institute`

`r PI_organization`

`r PI_address`

\
\
**QBiC contacts:**

***`r PM`***

`r PM_email`

Auf der Morgenstelle 10, 72076 Tuebingen


***

# Introduction and Aims

`r description`

```{r setup, echo=FALSE}
library(DT)
# condition: if there is a RIN column in Sample_preparations.tsv, prints a statement and the table with the RIN column; if there is no RIN column, or the column has one or more NAs, print only the metadata sheet.
QAfile = paste0(wd, "/DESeq2/metadata/metadata.tsv")
no_RIN = FALSE
RIN=FALSE
if(file.exists(QAfile)){
  df_QA <- read.table(QAfile, header=TRUE, sep="\t")
  if ("RIN" %in% colnames(df_QA)){
    if (any(is.na(df_QA$RIN))){
      no_RIN <- TRUE
    } else {
      RIN <- TRUE
    }
  } else {
    no_RIN <- TRUE
  }
} else {
  no_RIN <- TRUE
}
```

```{r RIN_block, echo=FALSE, results='asis', eval=RIN}
cat("***
# Sample metadata
The sample information, including quality assessment of the RNA samples, is summed up in the table below and can be found [here](./DESeq2/metadata/Sample_preparations.tsv). The RNA Integrity Number (RIN) value has a maximum value of 10.
\
")

datatable(df_QA, caption = "Quality Assessment", extensions = c('Buttons', 'FixedColumns'), "Autofill", options = list(dom = 'Bfrtip', scrollX = T, scrollCollapse = T, buttons = c('copy', 'csv', 'excel', 'pdf', 'print')))
```

```{r noRIN_block, echo=FALSE, results='asis', eval=no_RIN}
cat("***
# Sample metadata
The sample information is summed up in the table below.
\
")

if ("RIN" %in% colnames(df_QA)){
  df_QA$RIN <- NULL
}

datatable(df_QA, caption = "Sample metadata", extensions = c('Buttons', 'FixedColumns'), "Autofill", options = list(dom = 'Bfrtip', scrollX = T, scrollCollapse = T, buttons = c('copy', 'csv', 'excel', 'pdf', 'print')))
```

***

# Bioinformatics Analysis

```{r include=FALSE}
# loading this library produces an output, exlude with include=FALSE
library(tidyverse)
```

## Bioinformatics Analysis Quality Control Summary
Quality control of the data was performed using *FASTQC*. Details on the quality control parameters are available in the `MultiQC` report of all samples available [here](`r paste0(wd,"/QC/multiqc_report.html")`).

```{r echo=FALSE, results='asis'}
if(file.exists(paste0(wd,"/QC/fastqc.zip"))){
  cat("`FASTQC` reports for individual samples are additionally available [here](./QC/fastqc.zip).")
}
```

```{r echo=FALSE, results='asis'}
library(DT)
df_DE <- read.table(file = paste0(wd,"/QC/multiqc_data/multiqc_general_stats.txt"), header=TRUE, sep="\t")
is.num <- sapply(df_DE, is.numeric)
df_DE[is.num] <- lapply(df_DE[is.num], round, 2)
columns <- c("Sample", "featureCounts_mqc.generalstats.featurecounts.percent_assigned",
             "STAR_mqc.generalstats.star.uniquely_mapped_percent", "Cutadapt_mqc.generalstats.cutadapt.percent_trimmed", "FastQC_mqc.generalstats.fastqc.percent_duplicates", "FastQC_mqc.generalstats.fastqc.percent_gc",
             "FastQC_mqc.generalstats.fastqc.total_sequences")
df_DE_selected = df_DE %>% select(one_of(columns))
n_rows = nrow(df_DE_selected)
metadata <- df_DE <- read.table((paste0(wd, "/DESeq2/metadata/metadata.tsv")), header=TRUE, sep="\t")
metadata = metadata[,c("QBiC.Code","Secondary.Name")]
metadata$Sample = metadata$QBiC.Code
metadata$QBiC.Code = NULL
table_complete = merge(metadata,df_DE_selected,by="Sample")
write.table(table_complete,"./qc_summary.tsv", append = FALSE, quote = FALSE, sep = "\t",eol = "\n", na = "NA", dec = ".", row.names = F,  col.names = c("Sample Name", "Secondary Name", "% Assigned", "% Aligned", "% Trimmed", "% Dups", "% GC", "Seqs"), qmethod = c("escape", "double"))
if (nrow(table_complete) < 20){
cat("The table below (extracted from the `MultiQC` report) shows a summary of the bioinformatics analysis quality control including the percentage of assigned reads (% Assigned), the percentage of aligned reads (% Aligned), trimmed reads (% Trimmed), and duplicate reads (% Dups). The mean GC content (%GC) for each of the sequences and total number of reads (Seqs) is additionally indicated.")
datatable(table_complete, caption = "Quality control summary", colnames= c("Sample Name", "Secondary Name", "% Assigned", "% Aligned", "% Trimmed", "% Dups", "% GC", "Seqs"), extensions = c('Buttons', 'FixedColumns'), "Autofill", options = list(dom = 'Bfrtip', scrollX = T, scrollCollapse = T, buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), pageLength = 10))} else {
cat("The table downloadable under [this link](./QC/qc_summary.tsv) (extracted from the `MultiQC` report) shows a summary of the bioinformatics analysis quality control including the percentage of assigned reads (% Assigned), the percentage of aligned reads (% Aligned), trimmed reads (% Trimmed), and duplicate reads (% Dups). The mean GC content (%GC) for each of the sequences and total number of reads (Seqs) is additionally indicated.")
}
```

## Mapping Statistics

The alignment scores for each sample are shown below.

### Generalities about mapping rate {.tabset}

The mapping rate of data originating from the sequencing machine depends on several factors:

* The error rate of an Illumina sequencing machine is ~0.1%.
* The error rate for sequencing of GC rich regions is higher.
* Sequencing devices are not able to sequence highly repeated regions.
* Low quality of sequencing data results in a low mapping rate.
* There is no perfect algorithm to align the reads to a reference genome.
* There is no perfect transcript annotation of the reference genome.

#### Percercentages {-}
<center>
```{r, echo=F, out.width="160%", dpi=1800, fig.cap="STAR: Alignment Scores", fig.align='center'}
knitr::include_graphics(paste0(wd, "/QC/multiqc_plots/svg/mqc_star_alignment_plot_1_pc.svg"))
```

#### Number of reads {-}
```{r, echo=F, out.width="160%", dpi=1200, fig.cap="STAR: Alignment Scores", fig.align='center'}
knitr::include_graphics(paste0(wd, "/QC/multiqc_plots/svg/mqc_star_alignment_plot_1.svg"))
```
</center>
\

## Gene Counts {.tabset}
The gene assignments of featureCounts are shown below. Interpretation of the different assignment types:

**Unassigned_Ambiguity:** The reads could not be assigned because of its ambiguous character. This can be the case when e.g. a read maps to two different genes at the same time making it ambiguos. \
**Unassigned_MultiMapping:** The reads could not be assigned because they map to several distinct locations on the reference genome. \
**Unassigned_NoFeatures:** The reads could not be assigned because their mapping position in the reference genome is not annotated with any gene.

### Percercentages {-}
<center>
```{r, echo=F, out.width="160%", dpi=1200, fig.cap="featureCounts: Assignments", fig.align='center'}
knitr::include_graphics(paste0(wd, "/QC/multiqc_plots/svg/mqc_featureCounts_assignment_plot_1_pc.svg"))
```
</center>

### Number of reads {-}
<center>
```{r, echo=F, out.width="160%", dpi=1200, fig.cap="featureCounts: Assignments", fig.align='center'}
knitr::include_graphics(paste0(wd, "/QC/multiqc_plots/svg/mqc_featureCounts_assignment_plot_1.svg"))
```
</center>
\

## Read Distribution {.tabset}
The read distribution of mapped reads over different genome feature (like CDS exon, 5'UTR exon, 3' UTR exon, Intron, Intergenic regions) is shown below. The majority of the reads is assigned to CDS exons.

### Percercentages {-}
<center>
```{r, echo=F, out.width="160%", dpi=1200, fig.cap="RSeQC: Read Distribution", fig.align='center'}
knitr::include_graphics(paste0(wd, "/QC/multiqc_plots/svg/mqc_rseqc_read_distribution_plot_1_pc.svg"))
```
</center>

### Number of reads {-}
<center>
```{r, echo=F, out.width="160%", dpi=1200, fig.cap="RSeQC: Read Distribution", fig.align='center'}
knitr::include_graphics(paste0(wd, "/QC/multiqc_plots/svg/mqc_rseqc_read_distribution_plot_1.svg"))
```
</center>

***

# Exploratory Analysis

## Raw and normalized count tables

The raw count table and normalized count tables are available [here](./DESeq2/gene_counts_tables). 
The differential expression analysis is performed using the raw gene count table. 
For PCA analysis and heatmap plotting, the regularized logarithm (rlog) normalized gene counts were employed.

## Overview - First Exploration of the Data

### Principal Component Analysis (PCA) 
A PCA plot of the normalized expression values was created to visualize the overall effect of experimental variations and any batch effects among samples.

```{r, echo=F, out.width="2000px", dpi=1200, fig.cap="", fig.align='center'}
knitr::include_graphics(paste0(wd,"/DESeq2/plots/PCA_plot.svg"))
```

### Sample distance heatmap
A sample distance heatmap was generated from the distances of the normalized expression values for all samples. The closer the distance is to 0, the overall samples gene expression are more similar to each other.

```{r, echo=F, out.width="2000px", dpi=1200, fig.cap="", fig.align='center'}
knitr::include_graphics(paste0(wd,"/DESeq2/plots/Heatmaps_of_distances.svg"))
```

# Differential Expression analysis

## List of differentially expressed genes

Analysis of the differential gene expression was performed using `DESeq2`. 
Genes were considered differentially expressed (DE) when the adjusted p-value was lower than 0.05 (padj < 0.05). 
The adjusted p-value is calculated in the `DESeq2` package with the Benjamini-Hochberg method and 
helps reduce the number of false postives (not real differentially expressed genes). 
The model design in DESeq2 was:`r cat(paste(readLines(params$path_design), collapse=" "))`.
The list of the first 20 differentially expressed genes is shown below. 
The differentially expressed genes are marked with "DE" in the filter column. 
The full table, can be found [here](./DESeq2/results/final/final_list_DESeq2.tsv).

```{r}
library(DT)
library(tidyverse)
df_DE <- read.table(paste0(wd,"/DESeq2/final_gene_table/final_gene_list_DESeq2.tsv"), header=TRUE, sep="\t")
# list all the columns containing "padj"
padj = names(df_DE)[grepl("padj",names(df_DE))]
# take the last column in this list, order the dataframe according to ascending padj values in it; subset to the first 20 rows
df = head(df_DE %>% arrange(df_DE[[tail(padj, n=1)]]),20)
datatable(df, caption = "Outcome of Differential Expression analysis", extensions = c('Buttons', 'FixedColumns'), "Autofill", options = list(dom = 'Bfrtip', scrollX = T, scrollCollapse = T, buttons = c('copy', 'csv', 'excel', 'pdf', 'print')))
```

```{r setup2, echo=FALSE, message=FALSE, bootstrap.show.message=FALSE, include=FALSE}
library(DT)
contr=FALSE
df_contr <- try(read.table(params$path_contrasts, header=TRUE, sep="\t"))
if (!inherits(df_contr, 'try-error')){
  contrnames = grep('vs', colnames(df_contr), value=TRUE)
  contr <- TRUE
  }
```

```{r contrasts_block, echo=FALSE, results='asis', eval=contr}
cat("***
Additional tables for all individual contrasts are provided [here](./DESeq2/DE_genes_tables):
\
")

for (i in 1:length(contrnames)){
  cat(paste("* ", contrnames[i], "\n"))
}
```

## Plot normalized counts of the genes

Plots of the normalized counts were generated for some example genes. Count normalization was performed by sequencing depth and adding a pseudocount of 1/2 to allow for log scale plotting. The plots are available at [boxplots_example_genes](./DESeq2/plots/boxplots_example_genes).
```{r setup3, echo=FALSE, message=FALSE, bootstrap.show.message=FALSE, include=FALSE}
library(DT)
gene=FALSE
df_gene <- try(read.table(params$path_genelist))
if (!inherits(df_gene, 'try-error')){
  names = rownames(df_gene)
  gene = TRUE
}
```
```{r genes_block, echo=FALSE, results='asis', eval=gene}
cat("Plots for the [genes of interest](./DESeq2/metadata/gene_list.txt) can be found [here](./DESeq2/plots/boxplots_requested_genes/).
\
")
```

<!-- In case GO analysis was also included -->
```{r conditional_GO, echo=FALSE, results='asis', eval=config$GO}
version = packageVersion("clusterProfiler")
cat(paste("***
## Gene Ontology
Gene Ontology enrichment analysis was performed using `clusterProfiler v`", version, sep = ""))
cat(". GO terms that were enriched can be found in the respective table [here](./input/GO). The corresponding plot is shown below and can be found in the same folder:")
```

<center>
```{r, echo=F, out.width="2000px", dpi=1200, fig.cap="Enriched GO terms", fig.align='center', eval=config$GO}
knitr::include_graphics("./GO/GO_terms_enriched.png")
```
</center>

<!-- In case KEGG analysis was also included; differentiating the cases with and without contrasts -->
```{r conditional_KEGG_contrasts, echo=FALSE, results='asis', eval=config$KEGG}
cont = contrnames[1]
cat(paste0("## KEGG pathway analysis
Enriched KEGG pathways were found using a pathway analysis with `gProfileR`. The results are stored in [this folder](./gProfileR/), divided in subfolders, one for each contrast. Each subfolders contains a list of the enriched pathways, stored in the file ending with `results.tsv`, and also the plots of the KEGG pathway maps of the enriched pathways, that can be found in the subfolder named `KEGG_pathways`; in these maps, red corresponds to upregulated and green to downregulated parts of the pathways. There hardly is a difference in colour intensity between ~ 0.5 and 2.5 fold change.
Shown below you find the KEGG pathways enrichment plot for the example contrast ",contrnames[1],":",
"\"
)
```
```{r, out.width = "600px"}
knitr::include_graphics(paste0(wd,"/gProfileR/", 'DE_contrast_', contrnames[i], "/", 'DE_contrast_', contrnames[i], '_keg_pathway_enrichment_plot.png'))
```
```
# here a possible improvement for showing tabs: https://stackoverflow.com/questions/42631642/creating-dynamic-tabs-in-rmarkdown

#for (i in 1:length(contrnames)){
#  dirname = paste0('DE_contrast_', contrnames[i])
#  enriched_list = paste0(dirname, '_pathway_enrichment_results.tsv')
#  enriched_plot = paste0(dirname,'_keg_pathway_enrichment_plot.png')
#  path = paste0('/gProfileR/', dirname, '/', enriched_plot)
#  cat(paste0('### ',contrnames[i],' {-}
#  \
#  '))
#  knitr::include_graphics(paste0(wd,path))
#  cat("![](",(paste0(wd,path)),")")
#}
```

<!-- If a DEU analysis has also been performed -->
```{r conditional_DEU_results, echo=FALSE, results='asis', eval=config$DEU}
design = config$DEXSeq_design
cat(paste("## Differential Exon Usage
Analysis of the differential exon usage was performed using DEXSeq. Exons were considered differentially used (DEU) when the adjusted p-value was lower than 0.1 (padj < 0.1); the adjusted p-value helps reduce the number of false postives (not real differentially used exons). For each gene, the model design in DEXSeq had the formula: ", design, ". The list of all the differentially used exons can be found [here](./input/DEXSeq/DEXSeq_results_significant.txt).", sep=""))
```

```{r conditional_DEU_results2, echo=FALSE, results='asis', eval=config$DEU}
cat("### Fitted Expression Values Visualization
Plots of the fitted expression values of each exon were generated for a subset of 20 genes among those showing differential exon usage. The plots are available in [this folder](`r paste0(wd,/DESeq2/plots/boxplots_example_genes)`); in each graph both the experimental conditions are indicated, and the exon that showed significant differential usage is shown in purple.")
```

```{r conditional_outlook, echo=FALSE, results='asis', eval=!is.null(config$outlook)}
cat("***
# Outlook")
```
```{r conditional_outlook_content, echo=FALSE, results='asis', eval=!is.null(config$outlook)}
cat(config$outlook)
```

```{r, bootstrap.show.message=FALSE, echo=FALSE, message=FALSE}
# Import the .csv file output by the rna-seq Nextflow pipeline
version <- read.csv(file = params$path_versions, sep="\t", header = FALSE)
```
***
# Computational methods
<!--Different output in presence or absence of DEU analysis-->
Nextflow-based RNA-Seq pipeline release 1.3 https://github.com/nf-core/rnaseq [^1] was used for the bioinfomatics analysis. Aggregation of the bioinformatics workflow analysis was conducted by `MultiQC `r as.character(version[version$V1=="MultiQC",2])`` [^2] [@ewels2016multiqc]. `FASTQC `r as.character(version[version$V1=="FASTQC",2])`` [^3] [@andrews2010fastqc] was used to determine quality of the FASTQ files. Subsequently, adapter trimming was conducted with `Trim Galore `r as.character(version[version$V1=="Trim Galore!",2])`` [^4] [@krueger2012trim]. `STAR v`r substring(as.character(version[version$V1=="STAR",2]), 7, )`` [@Dobin2013] aligner was used to map the reads that passed the quality control to the human genome GRCh 38 (Ensembl release 92). The evaluation of the RNA-seq experiment was performed with `RSeQC `r as.character(version[version$V1=="RSeQC",2])`` [@wang2012rseqc] and read counting of the features (e.g. genes) with `featureCounts `r as.character(version[version$V1=="featureCounts",2])``[@liao2014featurecounts].
For differential expression analysis, the raw read count table resulting from `featureCounts` is processed with the R package `DESeq2 v`r packageVersion("DESeq2")`` [@love2014differential].
<!-- If a DEU analysis has also been performed -->
```{r conditional_DEU, echo=FALSE, results='asis', eval=config$DEU}
# version = packageVersion("DEXSeq")
# cat(paste("For differential exon usage, the table count produced by `featureCounts` is processed with the R package `DEXSeq v`",version," [@anders2012detecting].", sep=''))
```
Graphs were also produced in the `RStudio v1.1.456` with `r R.version.string` mainly using the `R` package `ggplot2 v`r packageVersion("ggplot2")``. Reports were produced using the `R` package `rmarkdown v`r packageVersion("rmarkdown")`` while `knitr v`r packageVersion("knitr")`` and `DT v`r packageVersion("DT")`` were running behind the scenes. Sample similarity heatmap was created using `edgeR `r as.character(version[version$V1=="edgeR",2])`` R package.

***
# Bibliography
[^1]: https://www.nextflow.io/
[^2]: http://multiqc.info/
[^3]: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/
[^4]: https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/
[^6]: http://smithlabresearch.org/software/preseq/
[^7]: https://ccb.jhu.edu/software/stringtie/
[^8]: http://broadinstitute.github.io/picard/
